### DISCUSSION TITLE
clear validation

### DISCUSSION THREAD

**Discussion by wdfinch** [react-hook-form/react-hook-form#4056] [Ideas]
Right now there's a helper called `clearErrors`. However, this helper doesn't seem to affect `formState.isValid`. I'm wondering if there's any way to manually clear validation without resetting or unregistering an input. My use case is I want to keep around form values, however, I don't' want them to affect validation if they are unmounted.

If there's no way to do this. I'm wondering what you all think of adding a new method called `clearValidation` or adding an option to `clearErrors` to also remove the validation for that input.

**Comment by bluebill1049** [comment]
is `unregister` the one you after?

**Reply by wdfinch** [reply]
According to the docs `unregister` removes the field's value from the form. I would like to keep the value, but not have it factored into validation. Does V7's `unregister` keep the value around?

**Reply by bluebill1049** [reply]
Hmm, no it removes everything.

**Reply by bluebill1049** [reply]
if you want to clear an error only, but still remain the value. i think this should be done at your validation rules, certain condition should remove error or else.

**Reply by wdfinch** [reply]
Ya just tested. 

Another idea I had was to do something like this.

```
 _.forOwn(formState.errors, (value, key) => {
   unregister(formState.errors[key])
 })
```

So in other words unregister the fields that have errors when I need to. The problem here is that `formState.errors` only populates once the validation `mode` criteria is met (I'm using `all`, so `onBlur` for me). So if the user hasn't touched the field, it won't be in there even if it's invalid.

I guess if there was some way to see the specific fields that were invalid, this could work also, however, it doesn't seem like this is part of `formState`. Could I access all the invalid fields somehow?

**Reply by wdfinch** [reply]
I thought about the above. My concern there was how the validation function would work when the component is unmounted. 

So for example, if I have an input and it uses some local variables in a component to determine if it's valid, how would the validation work when the component is unmounted and those variables are gone.

**Reply by wdfinch** [reply]
Just tested. This could work, but I think it would be better if the user could have more manual control over validation and/or visibility into which fields are valid/invalid.

**Reply by bluebill1049** [reply]
This is probably easier to achieve with schema validation, where each action will produce a result of schema result.

**Reply by wdfinch** [reply]
Thanks, I'll try that.

**Reply by bluebill1049** [reply]
in case you didn't know this: 

`unregister('test', { keepValid: true, keepErrors: true, keepTouched: true })`

**Reply by wdfinch** [reply]
Thanks. I guess for my use case I would need to the `keepValues` present in typescript but it looks like that's not an option

**Reply by bluebill1049** [reply]
if you want to keep the value, why would you unregister the input?

**Reply by wdfinch** [reply]
So I want the value but I don't want the value to be part of validation. So remove it from the formState.isValid

**Reply by bluebill1049** [reply]
This sounds like a validation condition, rather than to unregister the input itself.

**Reply by wdfinch** [reply]
Yes so going back to the original post, I guess all I'm saying is it would be nice to have more insight/control over validation. Right now all that is available is `formState.isValid`. 

For a number of use cases, it would be nice to see which fields are invalid and have some way to change the validation state. So in other words, remove certain fields from being validated (maybe when they are unmounted, but we don't want to use `unregister` and remove the value).
